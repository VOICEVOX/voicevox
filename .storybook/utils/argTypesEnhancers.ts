/**
 * Vue で emit されるイベントを Storybook の Controls に表示するための argTypesEnhancer
 * https://zenn.dev/shota_kamezawa/articles/36cd647264656c#storybook-%E3%81%AE%E8%A8%AD%E5%AE%9A
 */

import { toHandlerKey } from "vue";
import { type ArgTypesEnhancer, type StrictInputType } from "@storybook/types";

export const addActionsWithEmits: ArgTypesEnhancer = ({ argTypes }) => {
  const argTypesEntries = Object.entries(argTypes)
    .filter(([, argType]) => argType.table?.category === "events")
    .map(([name]) => {
      /**
       * 例：`click` という events に対して `onClick` という名称の argType + action を追加することで、
       * v-on によるイベントのバインディングが可能となる
       * @see https://ja.vuejs.org/guide/extras/render-function.html#v-on
       */
      const newName = toHandlerKey(name);
      const newArgType: StrictInputType = {
        name: newName,
        action: name,
        table: {
          disable: true, // Controls には表示しない
        },
      };

      return [newName, newArgType] as const;
    });

  return {
    ...argTypes,
    ...Object.fromEntries(argTypesEntries),
  };
};
