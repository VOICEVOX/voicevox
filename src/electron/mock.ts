/* eslint-disable @typescript-eslint/no-unused-vars */
import {
  defaultHotkeySettings,
  defaultToolbarButtonSetting,
  electronStoreSchema,
  EngineId,
  EngineInfo,
  engineSettingSchema,
  Sandbox,
  ThemeConf,
} from "@/type/preload";

declare const __availableThemes: ThemeConf[];

const storeName = "voicevox";

const engineInfos: EngineInfo[] = [
  {
    executionArgs: [],
    executionEnabled: false,
    executionFilePath: "",
    host: "http://127.0.0.1:50021",
    name: "VOICEVOX Engine",
    type: "default",
    uuid: EngineId("074fc39e-678b-4c13-8916-ffca8d505d1d"),
  },
];

const loadMock = () => {
  const electronMock: Sandbox = {
    isMock: true,
    async getAppInfos() {
      return {
        name: "VOICEVOX Web",
        version: "0.0.0",
      };
    },
    async getHowToUseText() {
      return "# How to use\ndummy";
    },
    async getPolicyText() {
      return "# Policy\ndummy";
    },
    async getOssLicenses() {
      return [];
    },
    async getUpdateInfos() {
      return [];
    },
    async getOssCommunityInfos() {
      return "";
    },
    async getQAndAText() {
      return "# Q&A\ndummy";
    },
    async getContactText() {
      return "# Contact\ndummy";
    },
    async getPrivacyPolicyText() {
      return "# Privacy Policy\ndummy";
    },
    async saveTempAudioFile(obj) {
      throw new Error(`Not implemented: saveTempAudioFile ${obj}`);
    },
    async loadTempFile() {
      throw new Error("Not implemented: loadTempFile");
    },
    async showAudioSaveDialog(obj) {
      throw new Error(`Not implemented: showAudioSaveDialog ${obj}`);
    },
    async showTextSaveDialog(obj) {
      throw new Error(`Not implemented: showTextSaveDialog ${obj}`);
    },
    async showVvppOpenDialog(obj) {
      throw new Error(`Not implemented: showVvppOpenDialog ${obj}`);
    },
    async showOpenDirectoryDialog(obj) {
      throw new Error(`Not implemented: showOpenDirectoryDialog ${obj}`);
    },
    async showProjectSaveDialog(obj) {
      throw new Error(`Not implemented: showProjectSaveDialog ${obj}`);
    },
    async showProjectLoadDialog(obj) {
      throw new Error(`Not implemented: showProjectLoadDialog ${obj}`);
    },
    async showMessageDialog(obj) {
      throw new Error(`Not implemented: showMessageDialog ${obj}`);
    },
    async showQuestionDialog(obj) {
      throw new Error(`Not implemented: showQuestionDialog ${obj}`);
    },
    async showImportFileDialog(obj) {
      throw new Error(`Not implemented: showImportFileDialog ${obj}`);
    },
    async writeFile(obj) {
      throw new Error(`Not implemented: writeFile ${obj}`);
    },
    async readFile(obj) {
      throw new Error(`Not implemented: readFile ${obj}`);
    },
    async openTextEditContextMenu() {
      throw new Error("Not implemented: openTextEditContextMenu");
    },
    async isAvailableGPUMode() {
      return false;
    },
    async isMaximizedWindow() {
      return false;
    },
    async onReceivedIPCMsg(channel, listener) {
      window.addEventListener("message", (event) => {
        if (event.data.channel === channel) {
          listener(event.data.args);
        }
      });
    },
    async closeWindow() {
      throw new Error("Not implemented: closeWindow");
    },
    async minimizeWindow() {
      throw new Error("Not implemented: minimizeWindow");
    },
    async maximizeWindow() {
      throw new Error("Not implemented: maximizeWindow");
    },
    async logError(...params) {
      console.error(...params);
    },
    async logWarn(...params) {
      console.warn(...params);
    },
    async logInfo(...params) {
      console.info(...params);
    },
    async engineInfos() {
      return engineInfos;
    },
    async restartEngine(engineId) {
      throw new Error(`Not implemented: restartEngine ${engineId}`);
    },
    async openEngineDirectory(engineId) {
      throw new Error(`Not implemented: openEngineDirectory ${engineId}`);
    },
    async hotkeySettings(newData) {
      if (newData !== undefined) {
        const hotkeySettings = await this.getSetting("hotkeySettings");
        const hotkeySetting = hotkeySettings.find(
          (hotkey) => hotkey.action == newData.action
        );
        if (hotkeySetting !== undefined) {
          hotkeySetting.combination = newData.combination;
        }
        await this.setSetting("hotkeySettings", hotkeySettings);
      }
      return this.getSetting("hotkeySettings");
    },
    async checkFileExists(file) {
      return false;
    },
    async changePinWindow() {
      throw new Error("Not implemented: changePinWindow");
    },
    async getDefaultHotkeySettings() {
      return defaultHotkeySettings;
    },
    async getDefaultToolbarSetting() {
      return defaultToolbarButtonSetting;
    },
    setNativeTheme(source) {
      // 何もしない
    },
    async theme(newData) {
      if (newData) {
        this.setSetting("currentTheme", newData);
        return;
      }
      return {
        currentTheme: await this.getSetting("currentTheme"),
        availableThemes: __availableThemes,
      };
    },
    vuexReady() {
      console.log("vuexReady");
    },
    getSetting(key) {
      const setting = electronStoreSchema.parse(
        JSON.parse(localStorage.getItem(storeName) || "{}")
      );
      // 同期でも使いたいので、async functionではなく手動でPromise.resolveを返す
      return Promise.resolve(setting[key]);
    },
    setSetting(key, newValue) {
      const setting = electronStoreSchema.parse(
        JSON.parse(localStorage.getItem(storeName) || "{}")
      );
      setting[key] = newValue;
      localStorage.setItem(storeName, JSON.stringify(setting));
      // 同期でも使いたいので、async functionではなく手動でPromise.resolveを返す
      return Promise.resolve(setting[key]);
    },
    async setEngineSetting(engineId, engineSetting) {
      await this.setSetting("engineSettings", {
        ...this.getSetting("engineSettings"),
        [engineId]: engineSetting,
      });
      return;
    },
    async installVvppEngine(path) {
      throw new Error(`Not implemented: installVvppEngine ${path}`);
    },
    async uninstallVvppEngine(engineId) {
      throw new Error(`Not implemented: uninstallVvppEngine ${engineId}`);
    },
    async validateEngineDir(engineDir) {
      throw new Error(`Not implemented: validateEngineDir ${engineDir}`);
    },
    async restartApp(obj) {
      window.location.reload();
    },
  };

  try {
    localStorage.setItem(
      storeName,
      JSON.stringify(
        electronStoreSchema.parse(
          JSON.parse(localStorage.getItem(storeName) || "{}")
        )
      )
    );
  } catch (e) {
    console.warn("Failed to load store, reset store");
    localStorage.setItem(
      storeName,
      JSON.stringify(electronStoreSchema.parse({}))
    );
  }

  const engineSettings = JSON.parse(
    localStorage.getItem("voicevox_engineSettings") || "{}"
  );
  for (const engineInfo of engineInfos) {
    if (!engineSettings[engineInfo.uuid]) {
      // 空オブジェクトをパースさせることで、デフォルト値を取得する
      engineSettings[engineInfo.uuid] = engineSettingSchema.parse({});
    }
  }
  electronMock.setSetting("engineSettings", engineSettings);

  // @ts-expect-error readonlyなので代入できないが、モックのため問題ない
  window.electron = electronMock;
};

export default loadMock;
