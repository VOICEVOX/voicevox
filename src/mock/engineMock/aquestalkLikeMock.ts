/**
 * AquesTalk 風記法テキストをパースするモジュール。
 * VOICEVOX ENGINEの voicevox_engine/tts_pipeline/kana_converter.py の移植。
 */

import { moraToPhonemes } from "./phonemeMock";
import { AccentPhrase, Mora } from "@/openapi";

enum ParseKanaErrorCode {
  UNKNOWN_TEXT = "判別できない読み仮名があります: {text}",
  ACCENT_TOP = "句頭にアクセントは置けません: {text}",
  ACCENT_TWICE = "1つのアクセント句に二つ以上のアクセントは置けません: {text}",
  ACCENT_NOTFOUND = "アクセントを指定していないアクセント句があります: {text}",
  EMPTY_PHRASE = "{position}番目のアクセント句が空白です",
  INTERROGATION_MARK_NOT_AT_END = "アクセント句末以外に「？」は置けません: {text}",
  INFINITE_LOOP = "処理時に無限ループになってしまいました...バグ報告をお願いします。",
}

const _LOOP_LIMIT = 300;

// AquesTalk 風記法特殊文字
const UNVOICE_SYMBOL = "_"; // 無声化
const ACCENT_SYMBOL = "'"; // アクセント位置
const NOPAUSE_DELIMITER = "/"; // ポーズ無しアクセント句境界
const PAUSE_DELIMITER = "、"; // ポーズ有りアクセント句境界
const WIDE_INTERROGATION_MARK = "？"; // 疑問形

// AquesTalk 風記法とモーラの対応。無声母音も含む。（音素長・音高 0 初期化）
const _kana2mora: Record<string, Mora> = {};
Object.entries(moraToPhonemes).forEach(([kana, [consonant, vowel]]) => {
  _kana2mora[kana] = {
    text: kana,
    consonant: consonant,
    consonantLength: consonant ? 0 : undefined,
    vowel: vowel,
    vowelLength: 0,
    pitch: 0,
  };

  if (["a", "i", "u", "e", "o"].includes(vowel)) {
    // 「`_` で無声化」の実装。例: "_ホ" -> "hO"
    // NOTE: 現行の型システムは Conditional Literal + upper に非対応.
    // FIXME: バリデーションする
    const upperVowel = vowel.toUpperCase();

    _kana2mora[UNVOICE_SYMBOL + kana] = {
      text: kana,
      consonant: consonant,
      consonantLength: consonant ? 0 : undefined,
      vowel: upperVowel,
      vowelLength: 0,
      pitch: 0,
    };
  }
});

/**
 * 単一アクセント句に相当するAquesTalk 風記法テキストからアクセント句オブジェクトを生成
 * longest matchによりモーラ化。入力長Nに対し計算量O(N^2)。
 */
function _textToAccentPhrase(phrase: string): AccentPhrase {
  // NOTE: ポーズと疑問形はこの関数内で処理しない

  let accentIndex: number | undefined = undefined;
  const moras: Mora[] = [];

  let baseIndex = 0; // パース開始位置。ここから右の文字列をstackに詰めていく。
  let stack = ""; // 保留中の文字列
  let matchedText: string | undefined = undefined; // 最後にマッチした仮名

  let outerLoop = 0;
  while (baseIndex < phrase.length) {
    outerLoop += 1;

    // 「`'` でアクセント位置」の実装
    if (phrase[baseIndex] === ACCENT_SYMBOL) {
      // 「アクセント位置はちょうど１つ」の実装
      if (moras.length === 0) {
        throw new Error(
          ParseKanaErrorCode.ACCENT_TOP.replace("{text}", phrase),
        );
      }
      if (accentIndex != undefined) {
        throw new Error(
          ParseKanaErrorCode.ACCENT_TWICE.replace("{text}", phrase),
        );
      }

      accentIndex = moras.length;
      baseIndex += 1;
      continue;
    }

    // モーラ探索
    // より長い要素からなるモーラが見つかれば上書き（longest match）
    // 例: phrase "キャ" -> "キ" 検出 -> "キャ" 検出/上書き -> Mora("キャ")
    for (let watchIndex = baseIndex; watchIndex < phrase.length; watchIndex++) {
      // アクセント位置特殊文字が来たら探索打ち切り
      if (phrase[watchIndex] === ACCENT_SYMBOL) {
        break;
      }
      stack += phrase[watchIndex];

      if (_kana2mora[stack]) {
        matchedText = stack;
      }
    }

    if (matchedText == undefined) {
      throw new Error(ParseKanaErrorCode.UNKNOWN_TEXT.replace("{text}", stack));
    } else {
      // push mora
      const baseMora = _kana2mora[matchedText];
      moras.push({ ...baseMora });

      baseIndex += matchedText.length;
      stack = "";
      matchedText = undefined;
    }

    if (outerLoop > _LOOP_LIMIT) {
      throw new Error(ParseKanaErrorCode.INFINITE_LOOP);
    }
  }

  if (accentIndex == undefined) {
    throw new Error(
      ParseKanaErrorCode.ACCENT_NOTFOUND.replace("{text}", phrase),
    );
  }

  return { moras, accent: accentIndex, pauseMora: undefined };
}

/**
 * AquesTalk 風記法テキストからアクセント句系列を生成
 */
export function parseKana(text: string): AccentPhrase[] {
  const parsedResults: AccentPhrase[] = [];
  if (text.length === 0) {
    throw new Error(ParseKanaErrorCode.EMPTY_PHRASE.replace("{position}", "1"));
  }

  let phraseBase = 0;
  for (let i = 0; i <= text.length; i++) {
    // アクセント句境界（`/`か`、`）の出現までインデックス進展
    if (
      i === text.length ||
      text[i] === PAUSE_DELIMITER ||
      text[i] === NOPAUSE_DELIMITER
    ) {
      let phrase = text.substring(phraseBase, i);
      if (phrase.length === 0) {
        throw new Error(
          ParseKanaErrorCode.EMPTY_PHRASE.replace(
            "{position}",
            String(parsedResults.length + 1),
          ),
        );
      }
      phraseBase = i + 1;

      // 「`？` で疑問文」の実装
      const isInterrogative = phrase.includes(WIDE_INTERROGATION_MARK);
      if (isInterrogative) {
        if (phrase.indexOf(WIDE_INTERROGATION_MARK) !== phrase.length - 1) {
          throw new Error(
            ParseKanaErrorCode.INTERROGATION_MARK_NOT_AT_END.replace(
              "{text}",
              phrase,
            ),
          );
        }
        // 疑問形はモーラでなくアクセント句属性で表現
        phrase = phrase.replace(WIDE_INTERROGATION_MARK, "");
      }

      const accentPhrase = _textToAccentPhrase(phrase);

      // 「`、` で無音付き区切り」の実装
      if (i < text.length && text[i] === PAUSE_DELIMITER) {
        accentPhrase.pauseMora = {
          text: "、",
          consonant: undefined,
          consonantLength: undefined,
          vowel: "pau",
          vowelLength: 0,
          pitch: 0,
        };
      }

      accentPhrase.isInterrogative = isInterrogative;

      parsedResults.push(accentPhrase);
    }
  }

  return parsedResults;
}
