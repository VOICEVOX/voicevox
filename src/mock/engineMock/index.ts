import { audioQueryToFrameAudioQueryMock } from "./audioQueryMock";
import { getEngineManifestMock } from "./manifestMock";
import {
  getSingersMock,
  getSpeakerInfoMock,
  getSpeakersMock,
} from "./characterResourceMock";
import { synthesisFrameAudioQueryMock } from "./synthesisMock";
import {
  replaceLengthMock,
  replacePitchMock,
  textToActtentPhrasesMock,
} from "./talkModelMock";
import {
  notesAndFramePhonemesAndPitchToVolumeMock,
  notesAndFramePhonemesToPitchMock,
  notesToFramePhonemesMock,
} from "./singModelMock";

import { cloneWithUnwrapProxy } from "@/helpers/cloneWithUnwrapProxy";
import {
  AccentPhrase,
  AccentPhrasesAccentPhrasesPostRequest,
  AudioQuery,
  AudioQueryAudioQueryPostRequest,
  DefaultApiInterface,
  EngineManifest,
  FrameAudioQuery,
  FrameSynthesisFrameSynthesisPostRequest,
  MoraDataMoraDataPostRequest,
  SingerInfoSingerInfoGetRequest,
  SingFrameAudioQuerySingFrameAudioQueryPostRequest,
  SingFrameVolumeSingFrameVolumePostRequest,
  Speaker,
  SpeakerInfo,
  SpeakerInfoSpeakerInfoGetRequest,
  SupportedDevicesInfo,
  SynthesisSynthesisPostRequest,
  UserDictWord,
} from "@/openapi";

/**
 * エンジンのOpenAPIの関数群のモック。
 * 実装されていない関数もある。
 */
export function createOpenAPIEngineMock(): DefaultApiInterface {
  const mockApi: Partial<DefaultApiInterface> = {
    async versionVersionGet(): Promise<string> {
      return "mock";
    },

    // メタ情報
    async engineManifestEngineManifestGet(): Promise<EngineManifest> {
      return getEngineManifestMock();
    },

    async supportedDevicesSupportedDevicesGet(): Promise<SupportedDevicesInfo> {
      return { cpu: true, cuda: false, dml: false };
    },

    // キャラクター情報
    async isInitializedSpeakerIsInitializedSpeakerGet(): Promise<boolean> {
      return true;
    },

    async initializeSpeakerInitializeSpeakerPost(): Promise<void> {
      return;
    },

    async speakersSpeakersGet(): Promise<Speaker[]> {
      return getSpeakersMock();
    },

    async speakerInfoSpeakerInfoGet(
      payload: SpeakerInfoSpeakerInfoGetRequest,
    ): Promise<SpeakerInfo> {
      return getSpeakerInfoMock(payload.speakerUuid);
    },

    async singersSingersGet(): Promise<Speaker[]> {
      return getSingersMock();
    },

    async singerInfoSingerInfoGet(
      paload: SingerInfoSingerInfoGetRequest,
    ): Promise<SpeakerInfo> {
      return getSpeakerInfoMock(paload.speakerUuid);
    },

    // トーク系
    async audioQueryAudioQueryPost(
      payload: AudioQueryAudioQueryPostRequest,
    ): Promise<AudioQuery> {
      const accentPhrases = await textToActtentPhrasesMock(
        payload.text,
        payload.speaker,
      );

      return {
        accentPhrases,
        speedScale: 1.0,
        pitchScale: 0,
        intonationScale: 1.0,
        volumeScale: 1.0,
        prePhonemeLength: 0.1,
        postPhonemeLength: 0.1,
        outputSamplingRate: getEngineManifestMock().defaultSamplingRate,
        outputStereo: false,
      };
    },

    async accentPhrasesAccentPhrasesPost(
      payload: AccentPhrasesAccentPhrasesPostRequest,
    ): Promise<AccentPhrase[]> {
      if (payload.isKana == true)
        throw new Error("AquesTalk風記法は未対応です");

      const accentPhrases = await textToActtentPhrasesMock(
        payload.text,
        payload.speaker,
      );
      return accentPhrases;
    },

    async moraDataMoraDataPost(
      payload: MoraDataMoraDataPostRequest,
    ): Promise<AccentPhrase[]> {
      const accentPhrase = cloneWithUnwrapProxy(payload.accentPhrase);
      replaceLengthMock(accentPhrase, payload.speaker);
      replacePitchMock(accentPhrase, payload.speaker);
      return accentPhrase;
    },

    async synthesisSynthesisPost(
      payload: SynthesisSynthesisPostRequest,
    ): Promise<Blob> {
      const frameAudioQuery = audioQueryToFrameAudioQueryMock(
        payload.audioQuery,
        {
          enableInterrogativeUpspeak:
            payload.enableInterrogativeUpspeak ?? false,
        },
      );
      const buffer = synthesisFrameAudioQueryMock(
        frameAudioQuery,
        payload.speaker,
      );
      return new Blob([buffer], { type: "audio/wav" });
    },

    // ソング系
    async singFrameAudioQuerySingFrameAudioQueryPost(
      payload: SingFrameAudioQuerySingFrameAudioQueryPostRequest,
    ): Promise<FrameAudioQuery> {
      const { score, speaker: styleId } = cloneWithUnwrapProxy(payload);

      const phonemes = notesToFramePhonemesMock(score.notes, styleId);
      const f0 = notesAndFramePhonemesToPitchMock(
        score.notes,
        phonemes,
        styleId,
      );
      const volume = notesAndFramePhonemesAndPitchToVolumeMock(
        score.notes,
        phonemes,
        f0,
        styleId,
      );

      return {
        f0,
        volume,
        phonemes,
        volumeScale: 1.0,
        outputSamplingRate: getEngineManifestMock().defaultSamplingRate,
        outputStereo: false,
      };
    },

    async singFrameVolumeSingFrameVolumePost(
      payload: SingFrameVolumeSingFrameVolumePostRequest,
    ): Promise<Array<number>> {
      const {
        speaker: styleId,
        bodySingFrameVolumeSingFrameVolumePost: { score, frameAudioQuery },
      } = cloneWithUnwrapProxy(payload);

      const volume = notesAndFramePhonemesAndPitchToVolumeMock(
        score.notes,
        frameAudioQuery.phonemes,
        frameAudioQuery.f0,
        styleId,
      );
      return volume;
    },

    async frameSynthesisFrameSynthesisPost(
      payload: FrameSynthesisFrameSynthesisPostRequest,
    ): Promise<Blob> {
      const { speaker: styleId, frameAudioQuery } =
        cloneWithUnwrapProxy(payload);
      const buffer = synthesisFrameAudioQueryMock(frameAudioQuery, styleId);
      return new Blob([buffer], { type: "audio/wav" });
    },

    // 辞書系
    async getUserDictWordsUserDictGet(): Promise<{
      [key: string]: UserDictWord;
    }> {
      // ダミーで空の辞書を返す
      return {};
    },
  };

  return mockApi satisfies Partial<DefaultApiInterface> as DefaultApiInterface;
}
