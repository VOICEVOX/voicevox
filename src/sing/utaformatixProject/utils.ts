import { Project as UfProject } from "@sevenc-nanashi/utaformatix-ts";
import { ExhaustiveError } from "@/type/utility";

export const singleFileProjectFormats = ["smf", "ufdata"] as const;
export type SingleFileProjectFormat = (typeof singleFileProjectFormats)[number];
export const multiFileProjectFormats = ["ust", "musicxml"] as const;
export type MultiFileProjectFormat = (typeof multiFileProjectFormats)[number];
export type ProjectFormat = SingleFileProjectFormat | MultiFileProjectFormat;

export const isSingleFileProjectFormat = (
  format: ProjectFormat,
): format is SingleFileProjectFormat =>
  singleFileProjectFormats.includes(format as SingleFileProjectFormat);

export const isMultiFileProjectFormat = (
  format: ProjectFormat,
): format is MultiFileProjectFormat =>
  multiFileProjectFormats.includes(format as MultiFileProjectFormat);

export const projectFileExtensions: Record<
  SingleFileProjectFormat | MultiFileProjectFormat,
  string
> = {
  smf: "mid",
  ufdata: "ufdata",
  ust: "ust",
  musicxml: "xml",
};

export const ufProjectToSingleFile = async (
  project: UfProject,
  format: SingleFileProjectFormat,
): Promise<Uint8Array> => {
  switch (format) {
    case "smf":
      return await project.toStandardMid();
    case "ufdata":
      return await project.toUfData();
    default:
      throw new ExhaustiveError(format);
  }
};

export const ufProjectToMultiFile = async (
  project: UfProject,
  format: MultiFileProjectFormat,
): Promise<Uint8Array[]> => {
  switch (format) {
    case "ust":
      return await project.toUst();
    case "musicxml":
      return await project.toMusicXml();
    default:
      throw new ExhaustiveError(format);
  }
};
