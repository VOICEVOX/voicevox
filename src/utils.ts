import si from "systeminformation";

interface ISysteminformationGraphicsControllers {
  model: string;
  vendor: string;
  bus: string;
  vram: number;
  vramDynamic: boolean;
}

interface ISysteminformationGraphicsDisplays {
  vendor: string;
  model: string;
  main: boolean;
  builtin: boolean;
  connection: string;
  sizeX: number;
  sizeY: number;
  pixelDepth: number;
  resolutionX: number;
  resolutionY: number;
  currentResX: number;
  currentResY: number;
  positionX: number;
  positionY: number;
  currentRefreshRate: number;
}

interface ISysteminformationGraphics {
  controllers: Array<ISysteminformationGraphicsControllers>;
  displays: Array<ISysteminformationGraphicsDisplays>;
}

export function detectNvidia(): Promise<boolean> {
  return si.graphics().then((data: ISysteminformationGraphics) => {
    return data.controllers.some(
      (a: ISysteminformationGraphicsControllers) =>
        a.vendor.indexOf("NVIDIA") !== -1 && a.vram >= 3072
    )
      ? true
      : false;
  });
}
