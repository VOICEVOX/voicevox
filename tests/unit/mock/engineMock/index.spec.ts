import { hash } from "../../utils";
import { createOpenAPIEngineMock } from "@/mock/engineMock";

describe("createOpenAPIEngineMock", () => {
  const mock = createOpenAPIEngineMock();

  it("versionVersionGet", async () => {
    const response = await mock.versionVersionGet();
    expect(response).toMatchSnapshot();
  });

  it("audioQueryAudioQueryPost", async () => {
    const response = await mock.audioQueryAudioQueryPost({
      text: "こんにちは",
      speaker: 0,
    });
    expect(response).toMatchSnapshot();
  });

  it("synthesisSynthesisPost", async () => {
    const audioQuery = await mock.audioQueryAudioQueryPost({
      text: "こんにちは",
      speaker: 0,
    });
    const response = await mock.synthesisSynthesisPost({
      audioQuery,
      speaker: 0,
    });
    expect(await hash(await response.arrayBuffer())).toMatchSnapshot();
  });

  it("singFrameAudioQuerySingFrameAudioQueryPost", async () => {
    const response = await mock.singFrameAudioQuerySingFrameAudioQueryPost({
      speaker: 0,
      score: {
        notes: [
          { id: "a", key: undefined, frameLength: 10, lyric: "" },
          { id: "b", key: 30, frameLength: 3, lyric: "て" },
          { id: "c", key: 30, frameLength: 3, lyric: "す" },
          { id: "d", key: 40, frameLength: 1, lyric: "と" },
          { id: "e", key: undefined, frameLength: 10, lyric: "" },
        ],
      },
    });
    expect(response).toMatchSnapshot();
  });

  it("frameSynthesisFrameSynthesisPost", async () => {
    const frameAudioQuery =
      await mock.singFrameAudioQuerySingFrameAudioQueryPost({
        speaker: 0,
        score: {
          notes: [
            { id: "a", key: undefined, frameLength: 10, lyric: "" },
            { id: "b", key: 30, frameLength: 3, lyric: "て" },
            { id: "c", key: 30, frameLength: 3, lyric: "す" },
            { id: "d", key: 40, frameLength: 1, lyric: "と" },
            { id: "e", key: undefined, frameLength: 10, lyric: "" },
          ],
        },
      });
    const response = await mock.frameSynthesisFrameSynthesisPost({
      frameAudioQuery,
      speaker: 0,
    });
    expect(await hash(await response.arrayBuffer())).toMatchSnapshot();
  });
});
