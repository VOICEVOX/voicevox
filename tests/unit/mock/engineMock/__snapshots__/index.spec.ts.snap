// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createOpenAPIEngineMock > audioQueryAudioQueryPost 1`] = `
{
  "accentPhrases": [
    {
      "accent": 5,
      "moras": [
        {
          "consonant": "k",
          "consonantLength": 0.0220625,
          "pitch": 4.7734375,
          "text": "コ",
          "vowel": "o",
          "vowelLength": 0.1140625,
        },
        {
          "consonant": undefined,
          "consonantLength": undefined,
          "pitch": 3.609375,
          "text": "ン",
          "vowel": "N",
          "vowelLength": 0.08312499999999999,
        },
        {
          "consonant": "n",
          "consonantLength": 0.022625,
          "pitch": 4.75,
          "text": "ニ",
          "vowel": "i",
          "vowelLength": 0.10843749999999999,
        },
        {
          "consonant": "ch",
          "consonantLength": 0.0400625,
          "pitch": 4.0234375,
          "text": "チ",
          "vowel": "i",
          "vowelLength": 0.10843749999999999,
        },
        {
          "consonant": "h",
          "consonantLength": 0.0315,
          "pitch": 4.796875,
          "text": "ハ",
          "vowel": "a",
          "vowelLength": 0.1509375,
        },
      ],
      "pauseMora": undefined,
    },
  ],
  "intonationScale": 1,
  "outputSamplingRate": 24000,
  "outputStereo": false,
  "pitchScale": 0,
  "postPhonemeLength": 0.1,
  "prePhonemeLength": 0.1,
  "speedScale": 1,
  "volumeScale": 1,
}
`;

exports[`createOpenAPIEngineMock > frameSynthesisFrameSynthesisPost 1`] = `"394cfbc01397e0b6fcc3433d9537aa850e6131f7d89048da6889e7375fe03a24"`;

exports[`createOpenAPIEngineMock > singFrameAudioQuerySingFrameAudioQueryPost 1`] = `
{
  "f0": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    46.17422889791779,
    46.08055781881967,
    46.08055781881967,
    46.16797823967303,
    46.18048040243588,
    46.18048040243588,
    82.27312267254713,
    82.21745071316357,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "outputSamplingRate": 24000,
  "outputStereo": false,
  "phonemes": [
    {
      "frameLength": 9,
      "noteId": "a",
      "phoneme": "pau",
    },
    {
      "frameLength": 1,
      "noteId": "b",
      "phoneme": "t",
    },
    {
      "frameLength": 2,
      "noteId": "b",
      "phoneme": "e",
    },
    {
      "frameLength": 1,
      "noteId": "c",
      "phoneme": "s",
    },
    {
      "frameLength": 2,
      "noteId": "c",
      "phoneme": "u",
    },
    {
      "frameLength": 1,
      "noteId": "d",
      "phoneme": "t",
    },
    {
      "frameLength": 1,
      "noteId": "d",
      "phoneme": "o",
    },
    {
      "frameLength": 10,
      "noteId": "e",
      "phoneme": "pau",
    },
  ],
  "volume": [
    0.7635414345273557,
    0.7635414345273557,
    0.7635414345273557,
    0.7635414345273557,
    0.7635414345273557,
    0.7635414345273557,
    0.7635414345273557,
    0.7635414345273557,
    0.7635414345273557,
    0.7732211475542967,
    0.7702900494608796,
    0.7702900494608796,
    0.7730257409255916,
    0.7734165541957456,
    0.7734165541957456,
    0.7737647610411076,
    0.7727873601766926,
    0.7635414345273557,
    0.7635414345273557,
    0.7635414345273557,
    0.7635414345273557,
    0.7635414345273557,
    0.7635414345273557,
    0.7635414345273557,
    0.7635414345273557,
    0.7635414345273557,
    0.7635414345273557,
  ],
  "volumeScale": 1,
}
`;

exports[`createOpenAPIEngineMock > synthesisSynthesisPost 1`] = `"23f4b910863418a7188648f7c5226a0f02b9d067b964be1690b69b1e9ffde7bc"`;

exports[`createOpenAPIEngineMock > versionVersionGet 1`] = `"mock"`;
